### DISCUSSION TITLE
Support PEP 582 (__pypackages__)

### DISCUSSION THREAD

**Discussion by NoelJacob** [microsoft/pylance-release#2584] [Enhancement]
Tools like [Pyflow](https://pypi.org/project/pyflow/) and [PDM](https://pypi.org/project/pdm/) are package managers using [Pep582 specification](https://www.python.org/dev/peps/pep-0582/#specification) to install packages. It'd be good if Pylance supported it.

**Comment by erictraut** [comment]
Thanks for the feature request.

Just a side note that mypy doesn't support PEP 582 currently either. See https://github.com/python/mypy/issues/10633.

**Comment by jakebailey** [comment]
This feels like a pretty unfortunate spec for analyzers... We already have issues resolving imports from scripts, and it seems like this PEP goes and makes it so that every script can have its own separate site-packages folder. It won't be as simple as asking python what the search paths are, because now there's a slew of paths that depend on which file is going to be read.

**Comment by frostming** [comment]
Note that PEP 582 doesn't mention the layout of `__pypackages__` directory, i.e. where should the scripts and libraries go, except for an example layout which only contains `lib` folder. If I read the PEP correctly, the libraries can reside in arbitrary deep sub-directories inside `__pypackages__`.

Kushal Das, the author of PEP 582, used a layout of `__pypackages__/lib/pythonX.Y/site-packages` in [his implementation](https://github.com/kushaldas/pep582/blob/1aebb799c3cf4585bf4be8ef4de116abcc698e2e/pep582.py#L46-L55), which is different from the example included in the PEP.

Therefore, I don't recommend starting the work to support PEP 582 in IDE before that point is clarified.

**Comment by DanLipsitt** [comment]
@kushaldas @zooba @dstufft @ncoghlan can any of you comment on the PEP 582 spec ambiguity above regarding directory layout?

Apologies if I have the handles of any of the PEP authors wrong.

**Comment by zooba** [comment]
The ambiguity is because the PEP is unfinished and has never been reviewed by anyone in a position to accept/approve it.

The _intent_ is that the "root" of a project (i.e. the directory that contains the script that is run directly) has a subdirectory that is added to sys.path. So it's one extra location relative to the main file, much like an explicitly configured PYTHONPATH value.

The feature creep led to people demanding separate directories based on Python version/tags, and then more directories for scripts and PATH entries. That's the point where I gave up on it, deciding that people couldn't possibly envision new tools and approaches for solving those issues and I couldn't be bothered trying to explain it. I have no idea what anyone else has done with the idea, but I'm not aware of any approved PEPs about it.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
